### PR TITLE
feat: [`Pay`] hooks and utils

### PR DESCRIPTION
**Pull request by 0xAlec**
**What changed? Why?**
- add hooks and utils for `Pay` component
- add `productId` support to `buildPayTransaction`
- remove `chainId`

**Notes to reviewers**

**How has it been tested?**
unit tests

### CODE REVIEW

File: src/pay/hooks/useCommerceContracts.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      });
+      const priceInUSDC = formatUnits(
+        BigInt(response.callData.feeAmount) +
+          BigInt(response.callData.recipientAmount),
+        6,
+      );
+
+      // Set insufficient balance flag, if applicable
+      userHasInsufficientBalanceRef.current =
+        Number.parseFloat(usdcBalance) < Number.parseFloat(priceInUSDC);
```
Comment (alessey) [inline src/pay/hooks/useCommerceContracts.tsx:58]: instead of setting a passed in ref, could this return hasInsufficientBalance | error?

**Review by alessey — APPROVED** [review]

File: src/pay/utils/handlePayRequest.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+export const handlePayRequest = async ({
+  address,
+  chargeHandler,
+  productId,
+}: {
+  address: Address;
+  chargeHandler?: () => Promise<string>;
+  productId?: string;
+}) => {
```
Comment (cpcramer) [inline src/pay/utils/handlePayRequest.ts:13]: `handlePayRequestReturn` type

File: src/pay/utils/getUSDCBalance.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import type { Config } from 'wagmi';
+import { readContract } from 'wagmi/actions';
+import { CONTRACT_METHODS, USDC_ADDRESS_BASE } from '../constants';
+
+export const getUSDCBalance = async ({
+  address,
+  config,
+}: {
+  address: Address;
+  config: Config;
```
Comment (cpcramer) [inline src/pay/utils/getUSDCBalance.ts:12]: type `getUSDCBalanceReturn`

File: src/api/mocks.ts [hunk 4]
```diff
@@ -28,6 +28,34 @@ export const MOCK_HYDRATE_CHARGE_SUCCESS_RESPONSE = {
     },
   },
 };
+export const MOCK_VALID_PRODUCT_ID = '1b03e80d-4e87-46fd-9772-422a1b693fb7';
+export const MOCK_CREATE_PRODUCT_CHARGE_SUCCESS_RESPONSE = {
```
Comment (cpcramer) [inline src/api/mocks.ts:32]: nice

File: src/api/types.ts [hunk 5]
```diff
@@ -18,8 +18,8 @@ export type APIError = {
  */
 export type BuildPayTransactionParams = {
   address: Address; // The address of the wallet paying
-  chainId: number; // The Chain ID of the payment Network (only Base is supported)
-  chargeId: string; // The ID of the Commerce Charge to be paid
+  chargeId?: string; // The ID of the Commerce Charge to be paid
+  productId?: string; // The ID of the product being paid for
```
Comment (cpcramer) [inline src/api/types.ts:22]: Curious if these should be required?

File: src/api/buildPayTransaction.test.ts [hunk 6]
```diff
@@ -1,26 +1,30 @@
-import { base, mainnet } from 'viem/chains';
 import { afterEach, describe, expect, it, vi } from 'vitest';
 import type { Mock } from 'vitest';
-import { PAY_HYDRATE_CHARGE } from '../network/definitions/pay';
+import {
+  CDP_CREATE_PRODUCT_CHARGE,
+  CDP_HYDRATE_CHARGE,
+} from '../network/definitions/pay';
```
Comment (cpcramer) [inline src/api/buildPayTransaction.test.ts:6]: Should we have these in a const file?

File: src/network/definitions/pay.ts [hunk 7]
```diff
@@ -1 +1,2 @@
-export const PAY_HYDRATE_CHARGE = 'pay_hydrateCharge';
+export const CDP_HYDRATE_CHARGE = 'cdp_hydrateCharge';
+export const CDP_CREATE_PRODUCT_CHARGE = 'cdp_createProductCharge';
```
Comment (cpcramer) [inline src/network/definitions/pay.ts:2]: Make a `constants.ts` file =?

Comment (cpcramer) [reply src/network/definitions/pay.ts:2]: Ah I see we have one, should these be added to it?

File: src/pay/utils/handlePayRequest.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+}: HandlePayRequestParams) => {
+  const buildPayTransactionParams: BuildPayTransactionParams = {
+    address,
+  };
+
+  if (chargeHandler) {
+    buildPayTransactionParams.chargeId = await chargeHandler();
+  } else if (productId) {
+    buildPayTransactionParams.productId = productId;
+  }
```
Comment (cpcramer) [inline src/pay/utils/handlePayRequest.ts:18]: Can we add a small comment here?

File: src/network/definitions/pay.ts (hunk 7 above)
Comment (0xAlec) [reply src/network/definitions/pay.ts:2]: actually reverted this, it seems like a breaking change and we shouldn't

Comment (0xAlec) [reply src/network/definitions/pay.ts:2]: lets stick with the `definitions/pay` file for now

File: src/api/types.ts (hunk 5 above)
Comment (0xAlec) [reply src/api/types.ts:22]: we'll add validation on the component itself

File: src/pay/utils/handlePayRequest.ts (hunk 8 above)
Comment (0xAlec) [reply src/pay/utils/handlePayRequest.ts:18]: will add the comment on the React component and not this util function

File: src/api/buildPayTransaction.ts [hunk 9]
```diff
@@ -1,37 +1,50 @@
-import { base } from 'viem/chains';
-import { PAY_HYDRATE_CHARGE } from '../network/definitions/pay';
-import { sendRequest } from '../network/request';
-import { PAY_UNSUPPORTED_CHAIN_ERROR_MESSAGE } from '../pay/constants';
+import {
+  CDP_CREATE_PRODUCT_CHARGE,
+  CDP_HYDRATE_CHARGE,
+} from '../network/definitions/pay';
+import { type JSONRPCResult, sendRequest } from '../network/request';
```
Comment (cpcramer) [inline src/api/buildPayTransaction.ts:5]: nit: we've been separating importing types. Unless this is what the formatter now prefers

File: src/pay/utils/handlePayRequest.ts (hunk 8 above)
Comment (0xAlec) [reply src/pay/utils/handlePayRequest.ts:18]: https://github.com/coinbase/onchainkit/pull/1338/files#diff-b7dd4bc7cc964af915e429bc56534f129be883d8ace8ace0523cc298b167965dR22-R24

File: src/api/buildPayTransaction.ts (hunk 9 above)
Comment (0xAlec) [reply src/api/buildPayTransaction.ts:5]: i just ran `yarn check`

File: src/api/buildPayTransaction.ts [hunk 10]
```diff
@@ -1,37 +1,50 @@
...
-      BuildPayTransactionResponse
-    >(PAY_HYDRATE_CHARGE, [
-      {
-        sender: address,
-        chainId: chainId,
-        chargeId,
-      },
-    ]);
+    let res: JSONRPCResult<BuildPayTransactionResponse>;
+    if (chargeId) {
```
Comment (cpcramer) [inline src/api/buildPayTransaction.ts:21]: I'm not sure when we should expect chargeId and productId, maybe add a comment?

Comment (0xAlec) [reply src/api/buildPayTransaction.ts:21]: they're two different methods and documented here - https://github.com/coinbase/onchainkit/pull/1338/files#diff-b7dd4bc7cc964af915e429bc56534f129be883d8ace8ace0523cc298b167965dR22-R24

Comment (0xAlec) [reply src/api/buildPayTransaction.ts:21]: improved the comment here though, but I don't think we should duplicate docstrings across helper/util functions

**Review by abcrane123 — APPROVED** [review]
lgtm ! nice job

**Pull request merged**